dwave/optimization/include/dwave-optimization/array.hpp

```diff
@@ -737,6 +738,27 @@ std::vector<ssize_t> broadcast_shape(const std::span<const ssize_t> lhs,
 std::vector<ssize_t> broadcast_shape(std::initializer_list<ssize_t> lhs,
                                      std::initializer_list<ssize_t> rhs);
 
+/// For "partial reduction", get the shape of the resulting shape of the array when preforming a
+/// reduction over an axis
+std::vector<ssize_t> partial_reduce_shape(const std::span<const ssize_t> array_shape,
+                                          const ssize_t axis);
+std::vector<ssize_t> partial_reduce_shape(std::initializer_list<ssize_t> input_shape,
+                                          const ssize_t axis);
+
+/// Gets the strides of a n-dimensional array assuming contiguous memory
+std::vector<ssize_t> as_contiguous_strides(const std::span<const ssize_t> shape);
+std::vector<ssize_t> as_contiguous_strides(std::initializer_list<ssize_t> shape);
+
+/// Convert a flat index to multi-index
+std::vector<ssize_t> unravel_index(const std::span<const ssize_t> strides, ssize_t index);
+
+/// Convert multi index to flat index
+ssize_t ravel_multi_index(const std::span<const ssize_t> strides,
+                          const std::span<const ssize_t> indices);
+
+ssize_t ravel_multi_index(const std::span<const ssize_t> strides,
+                          std::initializer_list<ssize_t> indices);
+
```

Comment:
For now I would have these live in `mathematical.cpp` alongside the node implementation. It doesn't look like we use them anywhere else. We can always promote them later if we want them.

Or do you think they would be insufficiently tested if we don't have access to them publicly?

Reply:
I put them here for exactly that purpose, as I was writing tests when developing. Since the broadcasting shape functions were in this file already, I thought it was consistent. I agree the tests are somewhat simple,  but better over-testing IMHO. If we find bugs, at least we can rule out they're not coming from raveling-unraveling indices.

Reply:
FWIW, IMO including `broad_shape()` in this file was a mistake for the reasons I outline above. But the testing point is reasonable.